File: tools/nebula-spark/src/main/java/com/vesoft/nebula/reader/AbstractNebulaIterator.java
/* Copyright (c) 2020 vesoft inc. All rights reserved.
 *
 * This source code is licensed under Apache 2.0 License,
 * attached with Common Clause Condition 1.0, found in the LICENSES directory.
 */

package com.vesoft.nebula.reader;

import com.facebook.thrift.TException;
import com.vesoft.nebula.bean.DataSourceConfig;
import com.vesoft.nebula.exception.GraphConnectException;
import com.vesoft.nebula.client.meta.MetaClientImpl;
import com.vesoft.nebula.client.storage.StorageClientImpl;
import com.vesoft.nebula.client.storage.processor.Processor;
import com.vesoft.nebula.data.Result;

import java.util.*;

import org.apache.spark.Partition;
import org.apache.spark.sql.Row;
import org.apache.spark.sql.RowFactory;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import scala.collection.AbstractIterator;

public abstract class AbstractNebulaIterator extends AbstractIterator<Row> {

    private static final Logger LOGGER = LoggerFactory.getLogger(AbstractNebulaIterator.class);

    protected Iterator<List<Object>> dataIterator;
    protected Iterator<Integer> scanPartIterator;

    protected List<List<Object>> resultValues = new ArrayList<>();
    protected StorageClientImpl storageClient;
    protected MetaClientImpl metaClient;
    protected Processor processor;

    protected Map<String, List<String>> returnCols;

    public AbstractNebulaIterator(Partition split,
                                  DataSourceConfig dataSourceConfig) {
        this.returnCols = dataSourceConfig.getReturnColMap();

        metaClient = new MetaClientImpl(dataSourceConfig.getHostAndPorts());
        try {
            metaClient.connect();
        } catch (TException e) {
            throw new GraphConnectException(e.getMessage(), e);
        }
        storageClient = new StorageClientImpl(metaClient);

        // allocate scanPart to this partition
        int totalPart = metaClient.getPartsAlloc(dataSourceConfig.getNameSpace()).size();
        NebulaPartition nebulaPartition = (NebulaPartition) split;
        List<Integer> scanParts = nebulaPartition.getScanParts(totalPart,
                dataSourceConfig.getPartitionNumber());
        LOGGER.info("partition index: {}, scanPart: {}", split.index(), scanParts.toString());
        scanPartIterator = scanParts.iterator();
    }

    @Override
    public abstract boolean hasNext();

    @Override
    public Row next() {
        return RowFactory.create(dataIterator.next().toArray());
    }

    protected abstract Iterator<List<Object>> process(Result result);
}

File: tools/nebula-spark/src/main/scala/com/vesoft/nebula/tools/connector/reader/AbstractNebulaIterator.scala
/* Copyright (c) 2020 vesoft inc. All rights reserved.
 *
 * This source code is licensed under Apache 2.0 License,
 * attached with Common Clause Condition 1.0, found in the LICENSES directory.
 */

package com.vesoft.nebula.tools.connector.reader

import java.util

import com.vesoft.nebula.client.meta.MetaClientImpl
import com.vesoft.nebula.client.storage.StorageClientImpl
import com.vesoft.nebula.client.storage.processor.Processor
import com.vesoft.nebula.data.{Property, Result}
import com.vesoft.nebula.tools.connector.NebulaUtils.NebulaValueGetter
import com.vesoft.nebula.tools.connector.{NebulaOptions, NebulaUtils}
import org.apache.spark.Partition
import org.apache.spark.sql.catalyst.InternalRow
import org.apache.spark.sql.catalyst.expressions.SpecificInternalRow
import org.apache.spark.sql.types.StructType

import org.slf4j.{Logger, LoggerFactory}

import scala.collection.mutable
import scala.collection.JavaConverters._

abstract class AbstractNebulaIterator extends Iterator[InternalRow] {
  private val LOG: Logger = LoggerFactory.getLogger(classOf[AbstractNebulaIterator])

  protected var dataIterator: Iterator[List[Property]] = _
  protected var scanPartIterator: Iterator[Integer]    = _
  protected var resultValues: mutable.ListBuffer[List[Property]] =
    mutable.ListBuffer[List[Property]]()

  protected var storageClient: StorageClientImpl                = _
  protected var metaClient: MetaClientImpl                      = _
  protected var processor: Processor[_]                         = _
  protected var returnCols: util.Map[String, util.List[String]] = new util.HashMap()

  private var schema: StructType = _

  def this(split: Partition, nebulaOptions: NebulaOptions, schema: StructType) {
    this()
    this.schema = schema

    nebulaOptions.getReturnColMap
      .foreach(entry => {
        this.returnCols.put(entry._1, entry._2.asJava)
      })

    this.metaClient = NebulaUtils.createMetaClient(nebulaOptions.getHostAndPorts, nebulaOptions)
    this.storageClient = NebulaUtils.createStorageClient(metaClient)

    // allocate scanPart to this partition
    val totalPart       = metaClient.getPartsAlloc(nebulaOptions.spaceName).size
    val nebulaPartition = split.asInstanceOf[NebulaPartition]
    val scanParts       = nebulaPartition.getScanParts(totalPart, nebulaOptions.partitionNums.toInt)
    LOG.info(s"partition index: ${split.index}, scanPart: ${scanParts.toString}")
    scanPartIterator = scanParts.iterator
  }

  override def hasNext: Boolean

  override def next(): InternalRow = {
    val getters: Array[NebulaValueGetter] = NebulaUtils.makeGetters(schema)
    val mutableRow                        = new SpecificInternalRow(schema.fields.map(x => x.dataType))

    val resultSet: Array[Property] = dataIterator.next().toArray
    for (i <- getters.indices) {
      getters(i).apply(resultSet(i), mutableRow, i)
      if (resultSet(i) == null) mutableRow.setNullAt(i)
    }
    mutableRow
  }

  protected def process(result: Result[com.vesoft.nebula.data.Row]): Iterator[List[Property]]
}
